[Wait for #3640] Feat/gradient checkpointing

yeonjae11 opened the pull request:
## Dependency of the PR
- Depends on: #3640 ([Fix] layer: Fix ActivationLayer input lifespan for non-inplace activations)

## Commits to be reviewed in this PR

<details><summary>[Feature] Add CheckpointBlock class and API declaration</summary><br />

Add gradient checkpointing infrastructure:
- CheckpointBlock class for managing checkpoint blocks
- addCheckpointBlock() API declaration in Model interface
- Build configuration for checkpoint_block

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped 

Co-authored-by: Hoyeon Jo <jhy213@snu.ac.kr>
Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

</details>

<details><summary>[Feature] Implement gradient checkpointing for memory optimization</summary><br />

Add complete gradient checkpointing implementation to reduce memory usage during training by recomputing activations during backward pass instead of storing them.

### Why Gradient Checkpointing?

Training large neural networks requires storing intermediate activations for backpropagation, which can consume significant GPU/CPU memory. Gradient checkpointing addresses this by:
- **Reducing peak memory usage by up to 38%** while maintaining training accuracy
- **Enabling training of larger models** within the same memory constraints
- **Trading minimal computation overhead (~10-15%)** for significant memory savings

### Key Implementation Changes:
- Extend TensorLifespan with RECOMPUTE variants for checkpoint tensors
- Add checkpoint support to LayerNode and LayerContext
- Implement NetworkGraph checkpoint block application and recomputation
- Update Manager for checkpoint-aware tensor allocation
- Extend ExecutionOrder to 5-tuple to include recompute phase

### Benchmark Results

**Memory Reduction (tested on transformer-style FFN models):**

| Model  | Layers | Normal Memory | Checkpoint Memory | Reduction |
|--------|--------|---------------|-------------------|-----------|
| GELU   | 16     | 150 MB        | 100 MB            | **33.3%** |
| GELU   | 32     | 282 MB        | 178 MB            | **36.9%** |
| GELU   | 64     | 431 MB        | 284 MB            | **34.1%** |
| SwiGLU | 16     | 150 MB        | 100 MB            | **33.3%** |
| SwiGLU | 32     | 282 MB        | 178 MB            | **36.9%** |
| SwiGLU | 64     | 545 MB        | 334 MB            | **38.7%** |

**Enables Training Larger Models (8GB memory limit):**

| Model  | Normal Mode      | Checkpoint Mode  | Improvement |
|--------|------------------|------------------|-------------|
| GELU   | 480M params (16L, 1024dim) | 600M params (20L, 1024dim) | **+25% larger model** |
| SwiGLU | 360M params (20L, 768dim)  | 480M params (16L, 1024dim) | **+33% larger model** |

**Latency Overhead:**
- Forward pass: Minimal overhead
- Backward pass: ~10-15% additional time due to recomputation
- Overall training time increase: ~8-14% (acceptable trade-off for memory savings)

### Correctness Verification ✅

**Training loss is identical between normal and checkpoint modes across all benchmark configurations.** This confirms that gradient checkpointing correctly recomputes activations and produces mathematically equivalent gradients.

| Configuration | Normal Loss | Checkpoint Loss | Match |
|---------------|-------------|-----------------|-------|
| GELU 16L      | 0.12842     | 0.12842         | ✅    |
| GELU 32L      | 0.10726     | 0.10726         | ✅    |
| SwiGLU 16L    | 0.14319     | 0.14319         | ✅    |
| SwiGLU 32L    | 0.10726     | 0.10726         | ✅    |

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped (unittest_nntrainer_exe_order excluded - requires update for 5-tuple ExecutionOrder)

Co-authored-by: Hoyeon Jo <jhy213@snu.ac.kr>
Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

</details>

<details><summary>[Docs] Add gradient checkpointing usage documentation</summary><br />

Add comprehensive documentation for gradient checkpointing feature:
- API usage examples with C++ code
- Performance characteristics and memory savings
- Implementation details and troubleshooting guide

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

</details>

### Summary

Gradient checkpointing is a widely-used memory optimization technique (adopted by PyTorch, TensorFlow, and other major frameworks) that enables training of larger models by trading computation for memory.

**API Usage:**
```cpp
model->addCheckpointBlock({"layer1", "layer2", "layer3"});  // Before initialization
```

### Key Benefits
1. **Memory Reduction**: Up to 38% reduction in peak memory usage
2. **Train Larger Models**: Enable training 25-33% larger models within same memory constraints  
3. **Minimal Overhead**: Only 10-15% latency increase
4. **Correctness Guaranteed**: Training loss identical to normal mode (verified across all benchmarks)

### Performance Visualization

#### Memory Savings with Minimal Latency Overhead
<img width="4920" height="2848" alt="gc_comparison_all" src="https://github.com/user-attachments/assets/6e6aae09-dda2-4dd4-92bd-5c211341d25f" />


#### Enables Training Larger Models
<img width="4179" height="2363" alt="max_trainable_by_memory_limit" src="https://github.com/user-attachments/assets/915c9499-6550-4bad-a0a1-e50af6a8b7f1" />


### Known Issues / TODO
- [ ] **unittest_nntrainer_exe_order**: Requires update for 5-tuple ExecutionOrder format. The test compares execution order output with golden data, which needs regeneration for the new format. (follow-up PR)
- [x] **Gradient checkpointing unit tests**: Add dedicated unit tests for:
  - CheckpointBlock class functionality
  - Recomputation correctness verification

Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

yeonjae11 committed:
[Fix] layer: Fix ActivationLayer input lifespan for non-inplace activations

ActivationLayer with supportInPlace()=false (e.g., swish, gelu) requires
the original input tensor during backward pass for gradient computation.
Previously, all ActivationLayer inputs were assigned FORWARD_FUNC_LIFESPAN,
causing the input to be overwritten before calcDerivative was called.

This fix:
- Add supportInPlace() virtual method to GraphNode base class
- Add override in LayerNode to expose layer's supportInPlace()
- Extend input lifespan to FORWARD_DERIV_LIFESPAN for ActivationLayers
  where supportInPlace() returns false

Resolves gradient computation errors in swish/gelu activations.

Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

yeonjae11 committed:
[Feature] Add CheckpointBlock class and API declaration

Add gradient checkpointing infrastructure:
- CheckpointBlock class for managing checkpoint blocks
- addCheckpointBlock() API declaration in Model interface
- Build configuration for checkpoint_block

Co-authored-by: Hoyeon Jo <jhy213@snu.ac.kr>
Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

yeonjae11 committed:
[Feature] Implement gradient checkpointing for memory optimization

Add complete gradient checkpointing implementation to reduce memory usage
during training by recomputing activations during backward pass instead of
storing them.

Key changes:
- Add CheckpointBlock class to define checkpoint boundaries
- Extend TensorLifespan with RECOMPUTE variants for checkpoint tensors
- Add checkpoint support to LayerNode and LayerContext
- Implement NetworkGraph checkpoint block application and recomputation
- Update Manager to handle checkpoint-aware tensor allocation
- Add NeuralNetwork API for checkpoint block configuration
- Extend ExecutionOrder to 5-tuple to include recompute phase

The implementation allows users to mark layer ranges as checkpoint blocks,
which will discard intermediate activations after forward pass and recompute
them during backward pass, trading computation for memory.

Co-authored-by: Hoyeon Jo <jhy213@snu.ac.kr>
Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

yeonjae11 committed:
[Docs] Add gradient checkpointing usage documentation

Add comprehensive documentation for gradient checkpointing feature:
- API usage examples with C++ code
- Multiple checkpoint block configuration
- Performance characteristics and memory savings
- When to use recommendations
- Implementation details and troubleshooting guide

Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

jhy213 committed:
[ unittest ] Add unittest for gradient checkpointing

Add initial gradient checkpointing unit coverage for model compilation and training paths. The tests exercise checkpoint block registration and recomputation behavior with representative graph layers.

Signed-off-by: jhy213 <jhy213@snu.ac.kr>

EunjuYang commented on nntrainer/graph/meson.build:
```diff
@@ -19,3 +21,10 @@ endforeach
+  nntrainer_inc += include_directories(gradient_checkpointing_test_dir)
+  nntrainer_sources += files(join_paths(gradient_checkpointing_test_dir,
+                                        'gradient_checkpointing_test_util.cpp'))
+endif
```
Instead of adding this test script here, how about adding `subdir(gradient_checkpointing)` to `test/unittest/mesonl.build` ?

jhy213 replied:
I implemented the tensor value comparison module as a test utility, but it is needed during forwarding() and backwarding() in network_graph.cpp, so I added it to nntrainer_inc and nntrainer_sources.

Would it be acceptable if I move the helper functions from gradient_checkpointing_test_util.cpp into network_graph.cpp and guard them with #ifdef ENABLE_TEST?

Thanks!

EunjuYang commented on test/unittest/gradient_checkpointing/unittest_gradient_checkpointing.cpp:
```diff
@@ -0,0 +1,412 @@
+ * @author Yeonjae Kim <duswo1120@snu.ac.kr>
+ * @author Hoyeon Jo <jhy213@snu.ac.kr>
+ * @bug No known bugs except for NYI items
+ */
```
What about adding negative test cases (e.g., a non‑consecutive block) or other edge cases, such as including only part of the network instead of entire layers?

jhy213 replied:
I added a testcase that checks non-consecutive block in compile phase.

EunjuYang reviewed (commented):
Great feature for reducing peak memory !
Please check the CI results and fix the uni-ttest failures :)

jhy213 committed:
[Fix] resolve unittest_nntrainer_exe_order failure

Update execution-order handling in the gradient checkpointing path so existing unittest_nntrainer_exe_order expectations remain valid after adding recompute execution slots.

Signed-off-by: jhy213 <jhy213@snu.ac.kr>

jhy213 committed:
[Fix] resolve Tizen GBS build on Ubuntu error

Adjust the gradient checkpointing build integration so Tizen GBS builds on Ubuntu can resolve the new graph source consistently across packaging configurations.

Signed-off-by: jhy213 <jhy213@snu.ac.kr>

jhy213 committed:
[Formatting] Apply Clang format and add description comment

Apply clang-format output and add missing description comments for the gradient checkpointing changes. This keeps style and documentation checks aligned with repository CI expectations.

Signed-off-by: jhy213 <jhy213@snu.ac.kr>

yeonjae11 committed:
Fix checkpoint block validation: use connection-based connectivity check

- Rename is_checkpoint_boundary to is_checkpoint_block_start for consistency
- Add input/output connection-based validation for checkpoint blocks
- Build reverse connection map from input connections (since output
  connections are not set before compile)
- Use bidirectional BFS to verify all layers form a single connected graph
- Correctly mark all start layers (no internal inputs) and end layers
  (no internal outputs) instead of just first/last in array
- Use block_layer_set for comparison to handle duplicates in layer_names

This addresses PR review comments:
1. Renamed boundary to start for naming consistency
2. Added validation to ensure checkpoint block layers are consecutive
   and form a connected subgraph

Signed-off-by: yeonjae11 <duswo1120@snu.ac.kr>

yeonjae11 commented on nntrainer/tensor/manager.cpp:
```diff
@@ -624,6 +642,13 @@ Manager::requestInputs(const GraphNode &node,
+  if (is_checkpoint_layer && !is_first_in_checkpoint_block) {
+    var_common_spec.ls = promoteToRecompute(var_common_spec.ls);
+  }
+
```
Fixed: Added connection-based validation using bidirectional BFS to ensure all layers in a checkpoint block form a single connected subgraph. If layers are not connected, an exception is thrown with details about disconnected layers.

jhy213 committed:
[ unittest ] Add testcase for gradient checkpointing

Add a negative validation case for checkpoint blocks that skip an intermediate layer. This covers invalid block connectivity so compile rejects non-consecutive checkpoint ranges.

Signed-off-by: jhy213 <jhy213@snu.ac.kr>

jhy213 committed:
[Formatting] Apply Clang format

Apply repository clang-format output to the gradient checkpointing graph and unit test sources. This keeps changed C++ lines aligned with nntrainer style checks before CI.

Signed-off-by: jhy213 <jhy213@snu.ac.kr>

jijoongmoon commented on nntrainer/graph/checkpoint_block.cpp:
```diff
@@ -0,0 +1,37 @@
+  }
+
+  start_layer = layers.front();
+  end_layer = layers.back();
```
It will be great if you add a check function whether the layers are consecutive.

jijoongmoon commented on nntrainer/graph/graph_node.h:
```diff
@@ -32,14 +32,16 @@ class GraphNode {
-   * 2. calcGradient
-   * 3. calcDerivative
-   * 4. ApplyGradient
+   * 2. Recompute Forward (for gradient checkpointing)
```
It is really great. However, we definitely need to check that every feature is working properly when we add this to the execution order. More importantly, we need to investigate the memory scheduling point more thoroughly.  

We can also compute the ideal peak memory when we enable this checkpointing through the memory scheduling, and we can also compare in an actual situation as well.

myungjoo commented:
```
unittest_gradient_checkpointing    FAIL
```

- When training=true, ***layer::forwarding() will update moving mean/variance (e.g., BatchNormalizationLayer::forwarding)
- At recompute of gradient checkpointing, ***layer::forawrding(..., training=true) is called ==> updated twice??
- I don't sure if this is really happening, but the unittest fail may be related. (if so, then, the unit test is designed pretty well! :D  )

jaemini-shin commented on test/unittest/gradient_checkpointing/gradient_checkpointing_test_util.h:
It'd be better to move these files under `nntrainer/graph`.

jaemini-shin commented on test/unittest/gradient_checkpointing/gradient_checkpointing_test_util.cpp:
```diff
@@ -0,0 +1,135 @@
+    }
+  }
+  EXPECT_TRUE(match);
+}
```
Please add a line here.

jaemini-shin reviewed (commented):
Looks like CI hasn't run for this PR, let's check it.

baek2sm committed:
Clean gradient checkpointing CI formatting

Remove markdown trailing whitespace and apply clang-format 14 to the rebased graph changes so static checks stay clean after the main rebase.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
Align checkpoint execution order type

Cast the iterator distance to the graph execution order type explicitly so the rebased checkpointing path stays warning-free on MSVC without changing runtime behavior.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
Stabilize checkpoint execution order users

Update existing graph and training paths to read the shifted execution order fields after adding the recompute phase. This keeps non-checkpointed training, cache flushing, and reinitialize bookkeeping aligned with the original operation ordering.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
Harden checkpoint block lifecycle

Tighten checkpoint block handling around application boundaries before pushing the rebased PR branch. The update rejects late registration, preserves checkpoint configuration during model copy and swap flows, and avoids compile-time output-connection assumptions when detecting checkpoint block endpoints.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
Fix checkpointing CI build isolation

Remove test-only verifier hooks from production graph dependencies and make debug validation aware of checkpoint initial tensors. This keeps Android package builds free of gtest while preserving targeted gradient checkpointing test coverage.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
Fix checkpointing test size calculation

Compute gradient checkpointing test buffer sizes with vector size_type before resizing so CodeQL does not flag integer multiplication widening.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
packaging: install checkpoint block header in debian

Add the gradient checkpointing public header to the Debian development package manifest so dh_missing sees the same installed files as Meson and the RPM spec.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

jijoongmoon reviewed (approved):
LGTM

baek2sm committed:
Fix gradient checkpointing crash, correctness, and bit-collision bugs

Null-check initial_inputs/initial_outputs in setBatch to prevent crash
when multi-input nodes have only some upstream layers checkpointed.
Reject BatchNormalization and Dropout layers in addCheckpointBlock since
their stateful forwarding corrupts the backward pass on recompute.
Move FORWARD_RECOMPUTE_LIFESPAN from 0b10000 to 0b1000000 to eliminate
spurious bit collision with EPOCH_LIFESPAN causing epoch tensors to absorb
the recompute exec order. Remove redundant double-nested if guard in the
needs_initial_outputs check and fix misleading getInput comment.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
test: add gradient checkpointing correctness and leak fixes

Fix memory leak in tests 01-03 by wrapping SimpleDataGenerator in
std::make_unique instead of raw new, preventing heap leaks on each run.
Add gradient_checkpointing_correctness_01 which trains two identical
FC-gelu-FC models with deterministic ones-initializer weights, one with
and one without addCheckpointBlock, and asserts final losses match within
1e-3 relative tolerance to guard recompute numerical correctness.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
layer_context: extend setBatch to resize initial_tensors on dynamic batch change

setBatch() already null-guards initial_inputs and initial_outputs but omits
initial_tensors. Add an equivalent null-guarded loop so that intermediate
tensors saved for gradient checkpointing recompute are correctly resized
when the batch size changes dynamically, preventing shape mismatches during
the recompute forward pass. Also add two compile-rejection tests that verify
BatchNormalization and Dropout layers are refused inside a checkpoint block.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
test: fix correctness_01 loss comparison for ARM Windows

MSE loss can be exactly 0.0 when the model fits training data perfectly;
this happens on ARM clang-cl where NEON GELU arithmetic rounds differently
from x64 AVX2. Replace the strict EXPECT_GT(loss_ref, 0.0f) with EXPECT_GE
since zero loss is a valid correct outcome. Use a symmetric denominator (max
of both losses plus epsilon) in the relative-error formula so neither loss
being 0 inflates the ratio and produces a spurious failure on platforms where
FP arithmetic rounds differently. The 1e-3 tolerance is unchanged and still
catches real recompute divergence which would be orders of magnitude larger.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm committed:
format: apply clang-format-14 to all branch-changed files

Run clang-format version 14 with style=file on every C/C++ file modified by this branch to satisfy cpp-linter CI requirements. No logic changes; formatting only.

Signed-off-by: SeungBaek Hong <sb92.hong@samsung.com>

baek2sm reviewed (approved):
LGTM
